www/js/lib/filecache.js

```diff
@@ -0,0 +1,187 @@
+     * @param {Uint16Array} value The value to store in the cache 
+     */
+    LRUCache.prototype.store = function (id, value) {
+        var entry = this._entries[id];
+        if (entry === undefined) {
+            entry = this._entries[id] = { id: id, prev: null, next: null, value: value };
+            this.insertAtTop(entry);
+            if (this._size >= this._limit) {
+                var e = this._last;
+                this.unlink(e);
+                delete this._entries[e.id];
+            } else {
+                this._size++;
+            }
+        } else {
+            entry.value = value;
+            this.moveToTop(entry);
+        }
+    };
+
+    /**
+     * Delete a cache entry
+     * @param {String} entry The entry to delete 
+     */
+    LRUCache.prototype.unlink = function (entry) {
+        if (entry.next === null) {
+            this._last = entry.prev;
+        } else {
+            entry.next.prev = entry.prev;
+        }
+        if (entry.prev === null) {
+            this._first = null;
+        } else {
+            entry.prev.next = entry.next;
+        }
+    };
+
+    /**
+     * Insert a cache entry at the top of the cache
+     * @param {String} entry The entry to insert 
+     */
+    LRUCache.prototype.insertAtTop = function (entry) {
+        if (this._first === null) {
+            this._first = this._last = entry;
+        } else {
+            this._first.prev = entry;
+            entry.next = this._first;
+            this._first = entry;
+        }
+    };
+
+    /**
+     * Move a cache entry to the top of the cache
+     * @param {String} entry The entry to move 
+     */
+    LRUCache.prototype.moveToTop = function (entry) {
+        this.unlink(entry);
+        this.insertAtTop(entry);
+    };
+
+    // Create a new cache
+    var cache = new LRUCache(MAX_CACHE_SIZE);
+
+    // Counters for reporting only
+    var hits = 0;
+    var misses = 0;
+
+    /**
+     * Read a certain byte range in the given file, breaking the range into chunks that go through the cache
+     * If a read of more than blocksize (bytes) is requested, do not use the cache
+     * @param {Object} file The requested ZIM archive to read from
+     * @param {Integer} begin The byte from which to start reading
+     * @param {Integer} end The byte at which to stop reading (end will not be read)
+     * @return {Promise<Uint8Array>} A Promise that resolves to the correctly concatenated data from the split ZIM file set
+     */
+    var read = function (file, begin, end) {
+        // Read large chunks bypassing the block cache because we would have to
+        // stitch together too many blocks and would clog the cache
+        if (end - begin > BLOCK_SIZE * 2) return file._readSplitSlice(begin, end);
+        var readRequests = [];
+        var blocks = {};
+        for (var i = Math.floor(begin / BLOCK_SIZE) * BLOCK_SIZE; i < end; i += BLOCK_SIZE) {
+            var block = cache.get(file.name + i);
```

Comment:
A bug might occur here in a **very** rare case : If I have 2 ZIM files named foo and foo1, and open them one after the other one (in the same session), there might be a mix if their cache blocks, for example between block 2 of foo1 and block 12 of foo : both would have the same id foo12.
I see at least 2 options to avoid that : 

- use a separator between the file name and the block number. But it needs to be a character that can never be used in a file name (on any OS)
- use a javascript object with 2 different properties in it, as the key. It's certainly safer and cleaner. It should be supported by using a https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map instead of a javascript Object. Based on this documentation from Mozilla, it also might be faster. And it seems to be supported by all our supported browsers (with some restrictions on IE11, but we probably don't use the unsupported functions)

Reply:
OK, thanks for the suggestion.

Reply:
@mossroy It seems that a new Map object is not easily extensible - https://stackoverflow.com/questions/52015305/how-to-extend-a-map-in-js . I could extend the prototype of the native Map object, but then every new Map (in this app) will have the defined functions -- would that be an issue? We currently use one other Map (a cache in app.js).

Alternatively, we could try to use principles of functional programming (#363), and not attach the functions to the prototype of the Cache itself. But I think then we would have some hybrid here of OOP and functional that is probably not ideal!

Any advice?

Reply:
I don't understand why you need to extend a Map.
What I was suggesting was to replace (line 48) `this._entries = {};` by `this._entries = new Map();`
and replace subsequent `_entries[id]` by `_entries.set(id, value)`or `_entries.get(id)` or `_entities.delete(id)`

Reply:
Ah, OK, thanks! I had understood that the entire FileCache object should be a Map with extended get, store and unlink functions in its prototype. What you say makes a lot more sense now. (I struggle with OOP.)

Reply:
I've managed to convert the code in `filecahce.js` to use a Map to store the file slices. However, the code is currently non-functional due to an issue with using a composite key that is an Array or an Object (I am using an array `key = [filename, id]`). While this stores correctly, to retrieve the data corresponding to this Array key, I would need to use the exact same array (in memory) as the retrieval key. It is not enough to make a `newkey = [filename, id]`, as even if the array key looks exactly the same, it is in fact a different array (i.e. a different location in memory), and `Map.get(newkey)` returns undefined, even though I can see that there is an entry in the Map using a key `[filename, id]`. This is the same reason that in console.log checking `{} === {}` returns `false`. The two Objects here are two different memory locations, so are not condered identical.

To use an Array (or Object) as a key, we would need to save a reference to a non-primitive instance of the Array, which is not viable for us. For discussion, see https://stackoverflow.com/questions/32660188/using-array-objects-as-key-for-es6-map

I still think using a Map is a good idea, but it looks like we are stuck with a composite key that is a String with a separator that would not be used in a ZIM filename. Rather than looking for exotic characters, I suggest we just use "@", which is most unlikely ever to appear at the end of a ZIM filename (especially since the filename part includes the `.zim` extension). Adding this would add approx 4K to the memory requirement (1 character * 4000 entries). I think we can be pretty sure no ZIM file will ever be called `wikipedia_en_all_maxi_2020-06.zim@1`, can't we?

An alternative would be to ensure the Cache is emptied on new ZIM load. I haven't checked whether in fact this is the case already.

Reply:
PS If we did empty the cache, we would no longer need to use the filename at all in the key. Since I modified peter-x's code to use absolute offsets from the start of the archive, not offsets relative to each ZIM file part, it's kind of redundant to use the filename in any case, unless we cannot empty the cache between ZIM loads for whatever reason. I'll investigate.

Reply:
I agree that it could be a good workaround
We don't really need to keep blocks of several ZIM files at the same time

Reply:
@mossroy This is now complete. To summarize: I've converted the storage to a Map, and have stripped out the use of the filename as part of the key. The key is now numeric (it is the absolute archive offset of the block to be read/stored). Since Maps have a size property, we no longer need to keep track of the Cache size with a counter. The Cache is reset in `zimFile.js` when a new ZIM is loaded. There is no need for explicit reset commands in `app.js`. I may be able to strip out the added generic filename in `zimFile.js` as well if it is no longer needed anywhere (will add a comment in code for that). But this is now passing tests and ready.